### DISCUSSION TITLE
When using useWindowVirtualizer across multiple screens, the scroll position is carried over. Is there a way to reset it?

### DISCUSSION THREAD

**Discussion by shunshimono** [TanStack/virtual#739] [Q&A]
Hello,

In my project, I am using Next.js and @tanstack/react-virtual. When using useWindowVirtualizer across multiple screens, the scroll position from the previous screen is automatically restored.

### Reproducible Environment
[CodeSandbox Example](https://codesandbox.io/p/devbox/virtualwindowlist-g99vpm?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxed44g700083b6jn4w60hvn%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxed44g600023b6j5wz18548%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxed44g700053b6j9es9sfii%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxed44g700073b6j04qs3bea%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxed44g600023b6j5wz18548%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxed44g600013b6j5tzq74q9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxed44g600023b6j5wz18548%2522%252C%2522activeTabId%2522%253A%2522clxed44g600013b6j5tzq74q9%2522%257D%252C%2522clxed44g700073b6j04qs3bea%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxed44g700063b6j1d3ft248%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252Fabout%2522%257D%255D%252C%2522id%2522%253A%2522clxed44g700073b6j04qs3bea%2522%252C%2522activeTabId%2522%253A%2522clxed44g700063b6j1d3ft248%2522%257D%252C%2522clxed44g700053b6j9es9sfii%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxed44g600033b6j7qb3120w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clxed44g600043b6jyfez4qiz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clxd8ukag000wdhgm6s7tfih9%2522%257D%255D%252C%2522id%2522%253A%2522clxed44g700053b6j9es9sfii%2522%252C%2522activeTabId%2522%253A%2522clxed44g600033b6j7qb3120w%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Reproduction Steps
Scroll on localhost:3000
Navigate to the ABOUT PAGE

### Expected Behavior
The scroll position should start from the top.

### Things I Tried but Didn't Work
Executing the measure method and initializing the cache on unmount: [measure method](https://tanstack.com/virtual/v3/docs/api/virtualizer#measure)
Setting initialOffset to 0: [initialOffset](https://tanstack.com/virtual/v3/docs/api/virtualizer#initialoffset)
Setting scrollToOffset to 0 on mount: [scrollToOffset](https://tanstack.com/virtual/v3/docs/api/virtualizer#scrolltooffset)

### Additional Observations
If I don't use measureElement, this issue doesn't occur.
I hypothesize that this is caused by the cache of measureElement being shared across screens. If anyone knows a solution, I would appreciate your support.

Thank you for reading.

**Comment by shunshimono** [comment]
memo:
- Even with history.scrollRestoration: 'manual', it still happens.
- It still happens with the latest version (v3.5.1).

**Comment by shunshimono** [comment]
Although I haven't fully traced the code yet, I have found the following points:

1. When scrolling on a screen that uses useWindowVirtualizer and measureElement, and then navigating to another screen using the same API via SPA, [resizeItem ](https://github.com/TanStack/virtual/blob/3562e0640d0c7dbd688d2de96d438135b90a2e17/packages/virtual-core/src/index.ts#L699)is triggered.
2. If shouldAdjustScrollPositionOnItemSizeChange is undefined, the [default scroll adjustment logic condition becomes true](https://github.com/TanStack/virtual/blob/3562e0640d0c7dbd688d2de96d438135b90a2e17/packages/virtual-core/src/index.ts#L710), causing the scroll position to revert to the previous screen's position.
3. By modifying the type definitions and the Virtualizer class to return false for shouldAdjustScrollPositionOnItemSizeChange, the issue can be resolved.


However, I find the functionality of adjusting the scroll position during resizing to be very useful. From my debugging, I have determined that the root cause is that this.scrollOffset or this.scrollAdjustments retains the previous values. I will continue reading the code to further investigate this issue.

**Comment by shunshimono** [comment]
[Update]
I have discovered that during page transitions in Next.js, the [scrollend ](https://github.com/TanStack/virtual/blob/3562e0640d0c7dbd688d2de96d438135b90a2e17/packages/virtual-core/src/index.ts#L167)event is triggered, causing the scroll position to be restored. This happens because the scroll option in Next.js's Link component is set to true by default, which attempts to adjust the scroll position before the transition.

I believe this can be addressed by custom controlling the timing of the observeElementOffset trigger.

**Comment by msmoter** [comment]
@shunshimono did you manage to fix this in your code? I am facing the same issue and nothing works properly. Sometimes `window.scrollTo(0,0)` during mounting works, but sometimes not and it does not satisfies me :)
